Projeto-3

AlanDouglasASouza opened the pull request:
Added terminal error handling.
https://github.com/terramagna/onboarding-rust/tree/main/essenciais/1-basico/10-projeto-3

AlanDouglasASouza committed:
chore: finish projeto-2

AlanDouglasASouza committed:
Merge branch 'projeto-2' into main

AlanDouglasASouza committed:
chore: added error handling

lffg commented on src/main.rs:
```diff
@@ -32,32 +35,61 @@ impl Terminal {
+
+impl TerminalError {
+
+    fn message_err(msg: TerminalError) -> String {
```
```suggestion
    fn message_err(self) -> String {
```

Dá para usar o `self` aqui! Aí, em vez de chamar `TerminalError::message_err(error)`, você pode fazer apenas `error.message_err()`.

AlanDouglasASouza replied:
Boa! Vou alterar isso.

lffg commented on src/main.rs:
```diff
@@ -32,32 +35,61 @@ impl Terminal {
+        Ok(())
+    }
+
+    fn show_error(&mut self, data: Result<(), TerminalError>) {
+
+        if let Err(error) = data {
+            println!("{}", TerminalError::message_err(error));
+        }
+       
+    }
```
Como a função `show_error` _sempre_ será usada para imprimir um erro na tela, aceitar um `Result` como parâmetro me parece desnecessário, WDYT?

```suggestion
    fn show_error(&mut self, error: TerminalError) {
            println!("{}", error.message_err());
    }
```

A vantagem de mudar a assinatura é que:
1. Você simplifica a implementação (afinal, sempre terá algo que representa um erro), de modo a poder remover o `if`.
2. Garante que só será possível chamar a função se você tiver um `TerminalError`, isto é, não poderá, por exemplo, passar um `Ok(())`.

AlanDouglasASouza replied:
Com certeza! Será implementado assim.

lffg commented on src/main.rs:
```diff
@@ -32,32 +35,61 @@ impl Terminal {
         buf.trim() == "s"
     }
 
     fn ask_for_new_todo(&mut self) -> Todo {          
```
Como a função `main` já está responsável por lidar com a "impressão" de erros na tela do usuário, o que acha de mudar a assinatura desta função para fazer com que ela retorne um `Result<Todo, TerminalError>`?

Isso também será vantajoso porque você também poderá fazer com que a função `should_ask_for_todo` retorne um `Result<bool, TerminalError>`, de modo que será possível utilizar o "operador" `?` para fazer o bubble up do erro. :)

AlanDouglasASouza replied:
Confesso que fiquei um pouco confuso em como implementar essa mudança. Subi as alterações e caso puder dar uma olhada onde poderia melhorar essa implementação agradeceria muito. Muito Obrigado!

lffg reviewed (commented):
Ficou legal! Gostei da ideia de ter criado um `TerminalError` :)

Apenas algumas sugestões para "padronizar" as assinaturas e _onde_ os erros são tratados. WDYT?

AlanDouglasASouza committed:
refac: modify Terminal error methods

Changed method signatures for clearer error handling feedback. If there is an error, an exit(1) was implemented.

luisholanda commented on src/main.rs:
```diff
@@ -30,34 +35,64 @@ impl Terminal {
-        self.stdin.read_line(&mut buf).unwrap();
 
-        buf.trim() == "s"
+        if let Err(msg_err) = self.stdin.read_line(&mut buf) {
+            self.show_error(TerminalError::StdoutErr(msg_err));
+            std::process::exit(1);         
+        }
+
+        Ok(buf.trim() == "s")
```
Aqui o melhor é você retornar o erro ao invés de finalizar o programa. Esse tipo de lógica normalmente fica em uma camada mais superior.

Uma forma de fazer isso é:

```suggestion
        if let Err(msg_err) = self.stdin.read_line(&mut buf) {
            Err(TerminalError::StdoutErr(msg_err))        
        } else {
            Ok(buf.trim() == "s")
        }
```

Mas também podemos usar [`Result::map_err`](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err) e `?`:

```rust
self.stdin.read_line(&mut buf).map_err(TerminalError::StdoutErr)?;

Ok(buf.trim() == "s")
```

Que acaba sendo a solução mais idiomática.

AlanDouglasASouza replied:
Entendi. Muito obrigado, me tirou uma duvida grande.

luisholanda commented on src/main.rs:
```diff
@@ -30,34 +35,64 @@ impl Terminal {
-        self.stdin.read_line(&mut new_todo).unwrap();
+        let mut new_todo = String::new();        
+
+        if let Err(msg_err) = self.stdin.read_line(&mut new_todo) {
+            self.show_error(TerminalError::StdoutErr(msg_err));
+            std::process::exit(1);
+        }
```
Mesma coisa aqui :)

AlanDouglasASouza replied:
Muito melhor, ja esta feito!! Muito obrigado.

luisholanda commented on src/main.rs:
```diff
@@ -1,13 +1,18 @@
-
-       ask_todo.show_todo(&todo);
+       
+       if let Ok(new_todo) = todo {
```
Hmmm, o que você fará se `ask_for_new_todo` retornar um erro?

AlanDouglasASouza replied:
Esqueci de tratar o erro nesse momento, mais refatorei usando match, me pareceu um pouco mais legível rsrs.
Valeeu!!

luisholanda commented on src/main.rs:
```diff
@@ -1,13 +1,18 @@
+       
+       if let Ok(new_todo) = todo {
+        if let Err(msg_err) = ask_todo.show_todo(&new_todo) {
+            Terminal::show_error(&mut ask_todo, TerminalError::WriteErr(msg_err))
```
Não seria melhor usar a sintaxe de método aqui?

AlanDouglasASouza replied:
Bem melhor! Obrigado!

luisholanda reviewed (commented):
Tá ficando bem legal!

AlanDouglasASouza committed:
refac: refactor the terminal error captures

Propagating errors and using method syntax in the main function.

AlanDouglasASouza commented on src/main.rs:
```diff
@@ -7,12 +7,15 @@ fn main() {
     loop {
        let mut ask_todo = Terminal::new();
        let todo = ask_todo.ask_for_new_todo();
```
Uma dúvida... Isso ou algo parecido funcionaria aqui?

``` let todo = ask_todo.ask_for_new_todo().map_err(ask_todo.show_error(TerminalError::WriteErr)); ```

luisholanda replied:
A única forma que conheço para fazer oq vc faz aqui é usando `match` mesmo.

AlanDouglasASouza replied:
Blz

luisholanda commented on src/main.rs:
```diff
@@ -30,34 +38,59 @@ impl Terminal {
-        
-        if !response {
+                
+        if let Ok(false) = self.should_ask_for_todo() {
```
Hmm, e o possível erro retornado aqui?
```suggestion
        if !self.should_ask_for_todo()? {
```

AlanDouglasASouza replied:
Show

luisholanda commented on src/main.rs:
```diff
@@ -7,7 +8,14 @@ fn main() {
-       ask_todo.show_todo(&todo);
+       match todo {
+           Ok(new_todo) => {            
+                if let Err(error) = ask_todo.show_todo(&new_todo) {
+                    ask_todo.show_error(TerminalError::WriteErr(error))
+                }        
```
Me é estranho o uso de `TerminalError` aqui, você não acha que seria melhor fazer `show_todo` já retornar um `TerminalError`?

AlanDouglasASouza replied:
Faz mais sentido mesmo

luisholanda commented on src/main.rs:
```diff
@@ -30,34 +38,59 @@ impl Terminal {
+                
+        if let Ok(false) = self.should_ask_for_todo() {
             println!("\nTodo list finalizado! 🤠");
             std::process::exit(0);            
```
Agora que você já viu o capítulo de enums, como você moveria essa parte para a main?

AlanDouglasASouza replied:
Eu mudei bastante pensando em usar essa logica de enum... Estou com duvida se atingi o ponto...

AlanDouglasASouza committed:
refac: refactor error handling

lffg commented on src/main.rs:
```diff
@@ -1,21 +1,35 @@
+            },
+            Ok(UserResponse::Other) => {
+                println!("\n🙁 Desculpe esse comando não é válido para esse processo...");
+                ()
```
Esse `()` que você colocou aqui se chama `unit`. Nesse caso, ele não é necessário porque toda expressão com um ponto e vírgula no final (como o `println!();` acima) _retorna_ um unit! :)

O unit representa o "valor nenhum" em Rust, como se fosse o `void` de outras linguagens. Não confundir com `null` ou `undefined`, que _são_ valores. O _unit_ é um valor sem massa :P õ-õ

AlanDouglasASouza replied:
Ficou sem peso nenhum mesmo, valeu!

lffg commented on src/main.rs:
```diff
@@ -1,21 +1,35 @@
+            Err(err) => ask_todo.show_error(err),            
+        }
     }
 }
```
Na função `main`, você chamou a função `show_error` várias vezes. Isso é _bom_ porque significa que você está tratando as várias fontes de erro do seu programa!

Todavia, ainda há espaço para melhoria, principalmente para evitar a repetição. Para isso, podemos tirar vantagem do operador `?`. <sup>Se você ainda não entendeu muito bem o que ele faz, tente me explicar aqui na thread com as suas palavras para consolidar o seu entendimento. (Se já entendeu, pode mandar também, se quiser)</sup>

Usando a `?`, um padrão muito comum é criar uma _outra_ função, ao lado da `main` (por exemplo, uma `fn run`) que retorna um `Result<(), TerminalError>`. Nesse caso, você pode utilizar a `?` várias vezes _dentro_ da função `run` para retornar os erros à `main`.

Assim, você só precisa ter o `if let Err(err) = run() { ... }` uma vez, dentro da função `main`. Todo o resto do código fica na `run`, WDYT?

A função `run` é necessária porque, se a `main` retornasse um `Result`, todos os erros retornados por ela seriam automaticamente convertidos num `panic!`, o que é justamente o que não queremos aqui, já que estamos tratando os erros com uma lógica customizada para _deixar_ de utilizar panics.

AlanDouglasASouza replied:
Boa, sobre o o operador ```?``` , o que eu entendi é que ele quando numa expressão que retorne um Result ele avalia se for um ```Err``` ele retorna da função inteira (como se fosse um ```return``` ) com o erro, já se for um ```Ok```, ele resolve o Result e continua executando a mesma função.
Até por isso a assinatura das funções em que usamos o ```?``` devem ser um Result<T, E>.

Foi o que eu entendi mais se eu estiver equivocado por favor me corrija :D

luisholanda replied:
É por aí mesmo @AlanDouglasASouza.

lffg reviewed (commented):
Está ficando realmente muito bom, parabéns! Só dois detalhes...

lffg commented on src/main.rs:
```diff
@@ -30,34 +52,56 @@ impl Terminal {
+    fn should_ask_for_todo(&mut self) -> Result<UserResponse, TerminalError> {
         let mut buf = String::new();
-        self.stdin.read_line(&mut buf).unwrap();
+        self.stdin.read_line(&mut buf).map_err(TerminalError::StdoutErr)?;
```
A seguir a lógica de separar `Stdout` e `Stdin`, aqui deveríamos usar a variante `StdinErr`, não?

AlanDouglasASouza replied:
Sim, estava passando despercebido disto. Obrigado!

AlanDouglasASouza committed:
refac: add run function

Moves the loop from the 'main' function to the 'run' function and takes advantage of the ? for error handling.

luisholanda reviewed (approved):
LGTM! Gostei muito da ideia do `UserResponse`!

BTW, seria interessante você formatar o código com `cargo fmt`.

AlanDouglasASouza commented:
Valeu, não tinha usado ainda ```cargo fmt``` muito pratico até pra subir em uma formatação unica e correta.
